### PR TITLE
Add configurable heartbeat

### PR DESCRIPTION
**Pull request by ricsiLT**
Allow configuring heartbeat. 

Added this quick mockup and quickly verified that connection is gone from RMQ GUI after inactivity. If this is not a good/full approach (I didn't test it with reliable consumer/producer) - feel free to update it. I'm not entirely sure how it should behave so didn't do much work on it. Just want to make sure that connections are correctly marked as dead since we have connections marked as "Active" for a week, and with heartbeat 0, I assume that server doesn't check whether they're really alive.

### CODE REVIEW

**Comment by Gsantomaggio** [comment]
@ricsiLT this is something that I wanted to change, so thank you.

I have to do some tests but adding a default value for the heartbeat makes sense

**Review by Gsantomaggio — CHANGES_REQUESTED** [review]
The PR is not completed yet.
The Heartbeat command is not handled:
```csharp
unknown or unexpected tag: 23 (Parameter 'tag')
   at RabbitMQ.Stream.Client.Client.HandleCorrelatedCommand(UInt16 tag, ReadOnlySequence`1& frame) in 
```

Working on it

**Comment by Gsantomaggio** [comment]
@ricsiLT added the heartbeat handler. If you have time can you please test it?

File: README.md [hunk 1]
```diff
@@ -411,6 +412,20 @@ You can use `MetadataHandler` to handle it:
...
+
+It is possible to configure the heart beat using:
+```csharp
+ var config = new StreamSystemConfig()
+{
+    Heartbeat = 30,
+}
+```
+- `60` seconds is the default value
+- `0` disable the heartbeat client side
```
[the remark this answers is not in the document]
Comment (Gsantomaggio) [reply README.md:426]: you are right.

File: RabbitMQ.Stream.Client/HeartBeatHandler.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+                _missedHeartbeat++;
+                LogEventSource.Log.LogWarning($"Heartbeat missed: {_missedHeartbeat}");
+                if (_missedHeartbeat <= 3)
+                {
+                    return;
+                }
+
+                // When the client does not receive the Heartbeat for three times the 
+                // client will be closed
+                LogEventSource.Log.LogWarning($"Too many Heartbeat missed: {_missedHeartbeat}");
```
Comment (ricsiLT) [inline RabbitMQ.Stream.Client/HeartBeatHandler.cs:47]: Since connection will be closed here, should this be `Critical`?

**Comment by ricsiLT** [comment]
Still testing but it's behaving strangely and maybe you can give me pointers:

Using:

* default producer (will test with reliable a bit later)
* heartbeat = 5
* publishing 100 messages to a fresh queue, 10s sleep between messages
* event listener attached so I can see logged messages
* RabbitMQ 3.9.15 (can upgrade if it's needed)

Once, the output was as follows:

```
Starting publishing
message: 0 - confirmed
message: 1 - confirmed
WARN: Heartbeat missed: 1
message: 2 - confirmed
message: 3 - confirmed
WARN: Heartbeat missed: 2
message: 4 - confirmed
message: 5 - confirmed
WARN: Heartbeat missed: 3
message: 6 - confirmed
WARN: Heartbeat missed: 4
WARN: Too many Heartbeat missed: 4
```

Connection was confirmed to be severed in GUI. No new messages were published or confirmed.

Next time I ran the test,

```
message: 0 - confirmed
message: 1 - confirmed
WARN: Heartbeat missed: 1
message: 2 - confirmed
message: 3 - confirmed
WARN: Heartbeat missed: 2
message: 4 - confirmed
message: 5 - confirmed
WARN: Heartbeat missed: 3
message: 6 - confirmed
message: 7 - confirmed
WARN: Heartbeat missed: 4
WARN: Too many Heartbeat missed: 4
message: 8 - confirmed
<...>
message: 51 - confirmed
WARN: Heartbeat missed: 1
message: 52 - confirmed
<...>
```

Not sure if it's a mistake on my part. Will test some more, just wanted to give some quick feedback.

File: RabbitMQ.Stream.Client/StreamSystem.cs [hunk 3]
```diff
@@ -15,6 +15,7 @@ public record StreamSystemConfig : INamedEntity
         public string UserName { get; set; } = "guest";
         public string Password { get; set; } = "guest";
         public string VirtualHost { get; set; } = "/";
+        public uint Heartbeat { get; set; } = 60;
```
Comment (ricsiLT) [inline RabbitMQ.Stream.Client/StreamSystem.cs:18]: Should this be changed to `TimeSpan` so it's more obvious what's passed in? Then `HeartBeatHandler` wouldn't need to have a variable named `timerIntervalSeconds`.

File: RabbitMQ.Stream.Client/HeartBeatHandler.cs (hunk 2 above)
Comment (Gsantomaggio) [reply RabbitMQ.Stream.Client/HeartBeatHandler.cs:47]: Will change to `LogEventSource.Log.LogError`

File: RabbitMQ.Stream.Client/StreamSystem.cs (hunk 3 above)
Comment (Gsantomaggio) [reply RabbitMQ.Stream.Client/StreamSystem.cs:18]: Good point, but `Heartbeat` is in seconds I would  leave it in seconds

**Comment by Gsantomaggio** [comment]
@[ricsiLT](https://github.com/ricsiLT) ready for another round. I changed the way to handle the interval

File: RabbitMQ.Stream.Client/StreamSystem.cs (hunk 3 above)
Comment (ricsiLT) [reply RabbitMQ.Stream.Client/StreamSystem.cs:18]: You could extract `TotalSeconds` from that `TimeSpan` internally but eh, it works.

**Comment by ricsiLT** [comment]
Seems good, don't see any missed heartbeats so far!

**Comment by ricsiLT** [comment]
Tested by running previously defined test + waiting for a long time on idle afterwards. Seen some missing heartbeats, but not more than 1 (should have added timestamps for future reference, but can confirm this was rare:

```
WARN: Heartbeat missed: 1
WARN: Heartbeat missed: 1
WARN: Heartbeat missed: 1
WARN: Heartbeat missed: 1
WARN: Heartbeat missed: 1
```

There's one more this which was strange, but I also got this in the logs:

```
WARN: Heartbeat missed: 0
```

Which doesn't seem possible from code? You incr before logging, so... found that interesting.

**Comment by Gsantomaggio** [comment]
>Which doesn't seem possible from code? You incr before logging, so... found that interesting.

That's possible the variable is changed by different threads. done via https://github.com/rabbitmq/rabbitmq-stream-dotnet-client/pull/130/commits/36ddc7051a19c059c5d7fb19099aadb2621d3dd6

File: RabbitMQ.Stream.Client/HeartBeatHandler.cs [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        // {
+        //     Heartbeat = 5,
+        // }
+        if (heartbeat > 0)
+        {
+            _timer.Enabled = true;
+            _timer.Interval = heartbeat * 1000;
+            _timer.Elapsed += (_, _) =>
+            {
+                sendHeartbeatFunc();
```
Comment (ricsiLT) [inline RabbitMQ.Stream.Client/HeartBeatHandler.cs:37]: `Rider` complains that this should probably be `await`ed.

**Comment by ricsiLT** [comment]
Not sure if you want to release it or waiting for test results. Built it and running in our staging for 2 hrs now, seems to behave nicely. In an hour we'll slow down message production to ~1 message/15min and see whether heartbeat keeps us alive (before, our network setup would kill the connection for idling).

**Comment by Gsantomaggio** [comment]
>Not sure if you want to release it or waiting for test results.

No rush; we can wait for the test results

> Built it and running in our staging for 2 hrs now, it seems to behave nicely.

Great :)

Thank you for helping @ricsiLT

**Comment by lukebakken** [comment]
@ricsiLT thanks for running a test in your environment with a very slow publisher.

**Review by lukebakken — APPROVED** [review]

**Review by Gsantomaggio — APPROVED** [review]
Remember to squash the commits during the merge!

**Comment by ricsiLT** [comment]
So, setup in our staging is +-:

* RabbitMQ 3.9.12 / Erlang 23.3.4.11
* reliable publisher
* 2 workers
* one worker will get a task +- every 15min

Tested for ~8 hours until errors happened, but I don't think they're related to heartbeat.

* Can confirm that heartbeat works in the sense that I do not see any missed HB and connections were properly closed on RMQ server side.
* At some point we start getting TimeoutErrors for any message we publish but I doubt that would be related to HB/disconnect from cluster - I don't see any socket exceptions or the like. The worst I see in logs is 
```
"WARN": "Semaphore Wait timeout during publishing."
```
* I'll test with 3.10.x version of RMQ on Monday only, not sure whether something would be different but we're due for upgrade anyways.

**Comment by ricsiLT** [comment]
Yup, investigated further - there were exceptions which look like misconfig somewhere - we are supplying LB but sometimes client decides to connect to localhost. In other words - unrelated to this :)

**Comment by Gsantomaggio** [comment]
Thank you @ricsiLT 
Let us know about the connections problem.

**Pull request merged**